Me: 2FA: Update backup codes list to allow codes to be copied to the clipboard

noxee opened the pull request:
Issue #385

This pull request seeks to add a button to the backup codes list that will allow a user to copy their codes to the clipboard in addition to being able to print them.

![chrome-copy-backup-codes](https://cloud.githubusercontent.com/assets/1409676/14373840/7d8a2a56-fd0b-11e5-9a8a-b97354c2d00e.gif)

**Implementation notes:**

This just uses the `ClipboardButton` component to provide the functionality of copying the codes to the clipboard.

**Testing instructions:**
1. Navigate to the [two-step authentication](calypso.localhost:3000/me/security/two-step) page.
2. Click the Generate New Backup Codes button.
3. Once new codes have been generated click the Copy button.
4. Confirm that the codes have been copied successfully by pasting clipboard into empty document (or wherever you see fit).

**Notes:**

There is currently a limitation in browsers that don't allow direct access to the clipboard (Safari, Safari iOS, etc) so the user will be prompted with a modal with an input box that will contain their codes and the can simply press CTRL+C (or other methods of copying) to achieve the same results. The only issue with this is that due to the input box being a single line the user won't be able to see all their codes.

![safari-copy-backup-codes](https://cloud.githubusercontent.com/assets/1409676/14373917/8eb0926a-fd0c-11e5-97ed-66ee80ce924a.gif)

cc @apeatling (let me know if there is someone else who I should be reviewing this PR)

timmyc commented:
I haven't checked out this branch yet, but will try to do so later today.  Just a thought though, with the addition of both the save and copy buttons, things might get a bit crowded - and additionally we could run into some overflow issues for translated strings.

Perhaps looking at using some gridicons for the copy and save buttons might be a good idea.  Such an example can be seen in stats:

<img width="733" alt="stats_ _trout_bummin_ _wordpress_com" src="https://cloud.githubusercontent.com/assets/22080/14716837/9f6e6572-07a3-11e6-9106-b71b29c1370b.png">

noxee commented:
Were you thinking of something like this? If we go with the gridicons for the copy and save button should the print button be updated as well?

![screen shot 2016-04-23 at 10 32 54 pm](https://cloud.githubusercontent.com/assets/1409676/14765536/6e121bf2-09a3-11e6-80c4-d3f4cb27762d.png)

timmyc commented:
@noxee exactly what I was thinking, but I'll label this PR as design review so a designer can chime in on this one.

rickybanister commented:
We have a print icon as well:

<img width="114" alt="screen shot 2016-04-25 at 12 36 09 pm" src="https://cloud.githubusercontent.com/assets/437258/14791108/56163de0-0ae2-11e6-8526-531b7f7d1c42.png">

noxee commented:
So this is what it looks like with Print, Copy, and Save as icons:

![screen shot 2016-04-25 at 6 16 36 pm](https://cloud.githubusercontent.com/assets/1409676/14802900/f74871fe-0b11-11e6-8452-b40aa4919084.png)

My only concern with having all icons is that it doesn't clearly indicate what the actions are (for most users they will be able to discern the meaning of each icon but for others it won't make any sense). So another option I thought was having a drop down list of actions like this:

![screen shot 2016-04-25 at 6 42 33 pm](https://cloud.githubusercontent.com/assets/1409676/14803332/c7ccf7d4-0b15-11e6-9bc3-dfa1ca417827.png)

One final option would be to use the icons but have a popover when the user hovers over the icon (although that ability won't be available to mobile users).

rickybanister commented:
The dropdown lacks an explicit action—it defaults to print, but in order to actually trigger the print dialog would you reselect print? Without a second somewhat superfluous button it would be confusing I think.

Adding the popovers to the icon-only buttons sounds just fine. Setting up 2FA entirely from a mobile device sounds a little bit advanced to begin with (you can't take a picture of your phone screen with your phone so you'd have to toggle the manual code entry, etc. Also printing and saving are interesting on phones anyway. 

I'm fine with the icon-only approach.

timmyc commented:
Agreed on @rickybanister's notes above.  As long as we have a tooltip for these icons, I think they will function well.  Yes it does pose a risk for mobile possibly, but would have to also agree with setting up 2FA via a mobile browser would be interesting regardless :)

noxee commented:
I've updated the Print and Copy buttons in this PR to use Gridicons rather than text. I didn't have any popover text to them so if you think that feature is need for the PR let me know.

timmyc commented:
@noxee I left some comments on #4637 that also affect this branch as well - specifically around how this looks on smaller viewports.

I think having a basic tooltip on both of these new buttons would be a very nice thing to have - perhaps:

"Copy codes to clipboard"

and

"Save codes to file"

drw158 commented:
Here are my suggestions for the "tooltip" labels

Print: `Print Codes`
Copy: `Copy Codes`
Download: `Download Codes`

"Download" because as a user I don't know what "save" means exactly.

We have a button group component that could be used for this to save space

https://wpcalypso.wordpress.com/devdocs/design/button-group

![screen shot 2016-06-17 at 3 53 06 pm](https://cloud.githubusercontent.com/assets/618551/16164413/a39092c4-34a3-11e6-8614-529cb1062ceb.png)

mtias commented:
I second @drw158 suggestion to look at the button-group component.

noxee commented:
Hi All,

Sorry for the delay in updating this PR, there was a holiday and disagreement between my laptop and glass of water thrown into the mix.

I've updated the buttons so that they are now in a ButtonGroup. This required me duplicating some code from the ClipboardButton component as the ButtonGroup only supports child elements of type Button.

I moved the Save/Download button code from the [PR 4637](https://github.com/Automattic/wp-calypso/pull/4637) to this branch as it seemed cumbersome to have to separate conversation regarding the same components.

I also used the tooltip text that @drw158 suggested.

Would it be worthwhile updating the Button component so that it supports having a tooltip? Or creating a new component that supports taking a Gridicon name and tooltip text?

The buttons still don't really stack that well on a mobile device as shown in the screenshot below. Should I do something to address this? e.g. collapse into a list?

![screen shot 2016-08-22 at 9 41 54 pm](https://cloud.githubusercontent.com/assets/1409676/17879809/115c50e4-68b2-11e6-9b66-0ebe4cb87594.png)

rickybanister commented:
I think it just requires some styles to prevent the buttons from wrapping down. Since they are icon-only they should be able to fit on a single line even with a narrow screen. It might require unsetting some styles that are triggered by the narrow viewport.

noxee commented:
@rickybanister I can make it so that the icons are together in one line like the screenshot below.

![screen shot 2016-08-23 at 6 42 57 pm](https://cloud.githubusercontent.com/assets/1409676/17914703/95f0aa6c-6961-11e6-80e1-c885aaa01b31.png)

Would it be better to change the order so that it looks something like this?

![image](https://cloud.githubusercontent.com/assets/1409676/17914750/f727c9fa-6961-11e6-826b-3644da7262a9.png)

rickybanister commented:
That looks good—I think they should stay above the 'All Finished' button since you must do something with the codes in order to wrap up.

noxee commented:
Cool, I've left it with the icons above the "All Finished" button.

rickybanister commented:
:+1

lancewillett commented:
Reviewing older pull requests, this looks close — can it be merged now?

lancewillett commented:
Closing as abandoned, no activity in 3+ months.

noxee commented:
@lancewillett I didn't abandon this. I was waiting for one of the Automattic devs to either approve this so that it could be merged or provide additional feedback.

lancewillett commented:
Cool, I didn't mean you had abandoned it personally — sorry if it felt that way. Thanks for the note. Let's reopen when devs have time to re-look at it again.

noxee commented:
Ah ok, yeah I did misconstrue what you said, thanks for the clarification. Let me know if there's anything else that needs to be done.

ebinnion commented on client/me/security-2fa-backup-codes-list/index.jsx:
```diff
@@ -215,16 +286,57 @@ module.exports = React.createClass( {
+						</Button>
+
+						<Button className="security-2fa-backup-codes-list__download"
+										disabled={ ! this.props.backupCodes.length }
```
There's extra indentation here. One tab please.

ebinnion commented on client/me/security-2fa-backup-codes-list/index.jsx:
```diff
@@ -215,16 +286,57 @@ module.exports = React.createClass( {
+						</Button>
+
+						<Button className="security-2fa-backup-codes-list__print"
+										disabled={ ! this.props.backupCodes.length }
```
Extra indentation here as well. One tab please.

a8ci18n commented on client/me/security-2fa-backup-codes-list/index.jsx:
```diff
@@ -213,19 +289,63 @@ module.exports = React.createClass( {
+								isVisible={ this.state.downloadCodesTooltip }
+								position="top"
+							>
+								{ this.translate( 'Download Codes' ) }
```
Alternate string suggestion: 
* `translate( 'Share embed codes and download links' )`&mdash; translations: **23**. 

[PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/pull/4603)

a8ci18n commented on client/me/security-2fa-backup-codes-list/index.jsx:
```diff
@@ -213,19 +289,63 @@ module.exports = React.createClass( {
+								isVisible={ this.state.printCodesTooltip }
+								position="top"
+							>
+								{ this.translate( 'Print Codes' ) }
```
Alternate string suggestion: 
* `translate( 'Print account recovery codes' )`&mdash; translations: **13**. 

[PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/pull/4603)

a8ci18n commented on client/me/security-2fa-backup-codes-list/index.jsx:
```diff
@@ -213,19 +289,63 @@ module.exports = React.createClass( {
+								isVisible={ this.state.copyCodesTooltip }
+								position="top"
+							>
+								{ this.translate( 'Copy Codes' ) }
```
Alternate string suggestion: 
* `translate( 'Copy' )`&mdash; translations: **82**. 

[PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/pull/4603)

noxee commented:
@ebinnion 

I've rebased with master and made the indentation changes.
Also when I was rebasing I squashed all the commits into one as it was getting quite messy with the multiple master merges I had done previously.

ebinnion commented on client/me/security-2fa-backup-codes-list/index.jsx:
```diff
@@ -162,7 +231,9 @@ module.exports = React.createClass( {
 	renderList: function() {
-		var backupCodes = this.props.backupCodes.length ? this.props.backupCodes : this.getPlaceholders();
+		const backupCodes = this.props.backupCodes.length
+							? this.props.backupCodes
```
There's still some excessive indentation here, but we can address this in a follow-up PR.

ebinnion commented:
Thanks for the contribution @noxee!
